src/data/list/defs.lean

```diff
@@ -170,13 +170,88 @@ def find (p : α → Prop) [decidable_pred p] : list α → option α
 def mfind {α} {m : Type → Type u} [monad m] [alternative m] (tac : α → m unit) : list α → m α :=
 list.mfirst $ λ a, tac a $> a
 
-def find_indexes_aux (p : α → Prop) [decidable_pred p] : list α → nat → list nat
-| []     n := []
-| (a::l) n := let t := find_indexes_aux l (succ n) in if p a then n :: t else t
+/-- Auxiliary definition for `foldl_with_index`. -/
+def foldl_with_index_aux (f : ℕ → α → β → α) : ℕ → α → list β → α
+| _ a [] := a
+| i a (b :: l) := foldl_with_index_aux (i + 1) (f i a b) l
+
+/-- Fold a list from left to right as with `foldl`, but the combining function
+also receives each element's index. -/
+def foldl_with_index (f : ℕ → α → β → α) (a : α) (l : list β) : α :=
+foldl_with_index_aux f 0 a l
+
+/-- Auxiliary definition for `foldr_with_index`. -/
+def foldr_with_index_aux (f : ℕ → α → β → β) : ℕ → β → list α → β
+| _ b [] := b
+| i b (a :: l) := f i a (foldr_with_index_aux (i + 1) b l)
+
+/-- Fold a list from right to left as with `foldr`, but the combining function
+also receives each element's index. -/
+def foldr_with_index (f : ℕ → α → β → β) (b : β) (l : list α) : β :=
+foldr_with_index_aux f 0 b l
+
+/-- `indexes l` is the list of indexes of `l`, i.e. `[0 .. l.length - 1]`. -/
+def indexes (l : list α) : list ℕ :=
+l.foldr_with_index (λ i _ is, i :: is) []
```

Comment:
Why even introduce this? It seems easier for people just to use `list.range (l.length)`, rather than giving an obscure construction of this list and then have to prove that it is the same.

Reply:
Removed.